File: scala-source/com/barrybecker4/puzzle/tantrix/solver/path/PathEvaluator.scala
// Copyright by Barry G. Becker, 2017. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix.solver.path

import com.barrybecker4.puzzle.tantrix.model.{HexTile, HexUtil, Tantrix}
import com.barrybecker4.puzzle.tantrix.solver.verification.{CompactnessCalculator, ConsistencyChecker, InnerSpaceDetector}
import PathEvaluator.*

/**
  * Evaluates the fitness of a tantrix path.
  * It gets the top score if it is a loop and all the path colors match.
  *
  * @author Barry Becker
  */
object PathEvaluator {
  /** If fitness is 0, then puzzle is solved. The worst fitness is this value. */
  val FITNESS_RANGE: Double = 10.0

  /** If more than this many tiles away from the required path length it doesn't matter */
  private val MAX_NUM_MISSING_TILES = 8

  private val UNPLACED_TILE_WEIGHT = 1.0

  /** How close are the endpoints of the primary path from forming a loop. */
  private val LOOP_PROXIMITY_WEIGHT = 0.4

  /** Weight to give if we actually have a primary path loop. */
  private val LOOP_WEIGHT = FITNESS_RANGE - 1.0

  /** Weight to give matching paths (includes secondary paths) */
  private val PATH_MATCH_WEIGHT = 0.7

  /** A measure of compactness. Avoids inner spaces. */
  private val COMPACTNESS_WEIGHT = 0.3

  private val INNER_SPACE_WEIGHT = 0.3
}

case class PathEvaluator() {

  /** The main criteria for quality of the path is
    * 0) How close the path is to using all the tiles.
    * 1) How close the ends of the path are to each other. Perfection achieved when we have a closed loop.
    * 2) Better if more matching secondary path colors
    * 3) Fewer inner spaces and a bbox with less area.
    * Getting a closed loop with no internal spaces is more important than having a lot of matching paths,
    * since tiles can be swapped to get more matches after the loop is determined.
    * @return a measure of how good the path is. Smaller number indicates better fitness.
    */
  def evaluateFitness(path: TantrixPath): Double = {
    val numTiles = path.size
    val isLoop = path.isLoop
    val consistencyChecker = ConsistencyChecker(path.tiles, path.primaryPathColor)
    val numFits = consistencyChecker.numFittingTiles
    val allFit = numFits == numTiles
    var numInnerSpaces = 0
    val consistentLoop = isLoop && allFit
    var perfectLoop = false
    if (consistentLoop) {
      val tantrix = new Tantrix(path.tiles)
      val innerDetector = InnerSpaceDetector(tantrix)
      numInnerSpaces = innerDetector.numInnerSpaces()
      perfectLoop = numInnerSpaces == 0 && numTiles == path.desiredLength
    }

    val fitness =
      if (perfectLoop) 0.0
      else if (isLoop && numTiles == path.desiredLength) calcFitnessForImperfectLoop(path, numFits, numInnerSpaces)
      else calcFitnessForNonLoop(path, numTiles, numFits)

    assert(fitness >= 0, "Fitness (" + fitness + ") must be >= 0")
    assert(fitness <= FITNESS_RANGE, "Fitness (" + fitness + ") must be <= " + FITNESS_RANGE)
    assert(!fitness.isNaN, "Invalid fitness  isLoop=" + isLoop + " consistentLoop=" + consistentLoop +
      " numTiles=" + numTiles)

    fitness
  }

  /**
    * Heuristic function to determine the fitness of an imperfect loop path
    * @return approximate fitness
    */
  private def calcFitnessForImperfectLoop(path: TantrixPath, numFits: Int, numInnerSpaces: Int): Double = {
    val distance = path.getEndPointDistance
    assert(distance == 0, "Distance from loop end points should be 0. " + distance)

    val compactness = CompactnessCalculator().determineCompactness(path)

    val sum: Double = LOOP_WEIGHT +
        PATH_MATCH_WEIGHT * (numFits.toDouble / path.desiredLength) +
        COMPACTNESS_WEIGHT * compactness - INNER_SPACE_WEIGHT * numInnerSpaces
    val score: Double = FITNESS_RANGE - sum  // between 0 and 1
    assert(score >= 0, "Score not positive. " + score)
    score
  }

  /**
    * Heuristic function to determine the fitness of a non-loop path
    * @return approximate fitness
    */
  private def calcFitnessForNonLoop(path: TantrixPath, numTiles: Int, numFits: Int): Double = {
    val distance = path.getEndPointDistance
    val compactness = CompactnessCalculator().determineCompactness(path)

    val sum: Double =
      UNPLACED_TILE_WEIGHT * (MAX_NUM_MISSING_TILES - Math.min(path.desiredLength - numTiles, MAX_NUM_MISSING_TILES)) +
      LOOP_PROXIMITY_WEIGHT * distance / numTiles +
      PATH_MATCH_WEIGHT * (numFits.toDouble / numTiles) +
      COMPACTNESS_WEIGHT * compactness
    val score: Double = FITNESS_RANGE - sum
    assert(score >= 0, "Score not positive. " + score)
    score
  }

}

File: source/com/barrybecker4/puzzle/tantrix1/solver/path/PathEvaluator.java
// Copyright by Barry G. Becker, 2012. Licensed under MIT License: http://www.opensource.org/licenses/MIT
package com.barrybecker4.puzzle.tantrix1.solver.path;

import com.barrybecker4.common.geometry.Location;
import com.barrybecker4.puzzle.tantrix1.model.HexTile;
import com.barrybecker4.puzzle.tantrix1.model.HexUtil;
import com.barrybecker4.puzzle.tantrix1.model.Tantrix;
import com.barrybecker4.puzzle.tantrix1.model.TilePlacement;
import com.barrybecker4.puzzle.tantrix1.solver.verification.ConsistencyChecker;
import com.barrybecker4.puzzle.tantrix1.solver.verification.InnerSpaceDetector;

import java.util.HashSet;
import java.util.Set;


/**
 * Evaluates the fitness of a tantrix path.
 * It gets the top score if it is a loop and all the path colors match.
 *
 * @author Barry Becker
 */
public class PathEvaluator {

    /** When reached, the puzzle is solved. */
    public static final double SOLVED_THRESH = 3.1;

    /** How close are the endpoints of the primary path from forming a loop. */
    private static final double LOOP_PROXIMITY_WEIGHT = 0.3;

    /** Weight to give if we actually have a primary path loop. */
    private static final double LOOP_WEIGHT = 0.3;

    /** Weight to give matching paths (includes secondary paths) */
    private static final double PATH_MATCH_WEIGHT = 1.0;

    /** We have a loop and all paths match */
    private static final double CONSISTENT_LOOP_BONUS = 0.6;

    /** consistent loop and no inner spaces. */
    private static final double PERFECT_LOOP_BONUS = 2.0;

    /** A measure of compactness. Avoids inner spaces. */
    private static final double COMPACTNESS = 0.2;


    /**
     * The main criteria for quality of the path is
     *  1) How close the ends of the path are to each other. Perfection achieved when we have a closed loop.
     *  2) Better if more matching secondary path colors
     *  3) Fewer inner spaces and a bbox with less area.
     * @return a measure of how good the path is.
     */
    public double evaluateFitness(TantrixPath path) {

        int numTiles = path.size();
        double distance = path.getEndPointDistance();
        boolean isLoop = distance == 0 && path.isLoop();

        ConsistencyChecker checker = new ConsistencyChecker(path.getTilePlacements(), path.getPrimaryPathColor());
        int numFits = checker.numFittingTiles();
        boolean allFit = (numFits == numTiles);
        boolean consistentLoop = isLoop && allFit;
        boolean perfectLoop = false;
        double compactness = determineCompactness(path);

        if (consistentLoop) {
            Tantrix tantrix = new Tantrix(path.getTilePlacements());
            InnerSpaceDetector innerDetector = new InnerSpaceDetector(tantrix);
            perfectLoop = !innerDetector.hasInnerSpaces();
        }
        assert numFits <= numTiles;

        double fitness = SOLVED_THRESH
                - LOOP_PROXIMITY_WEIGHT * (numTiles - distance) / (0.1 + numTiles)
                - (isLoop ? LOOP_WEIGHT : 0)
                - (double)numFits / numTiles * PATH_MATCH_WEIGHT
                - compactness * COMPACTNESS
                - (consistentLoop ? CONSISTENT_LOOP_BONUS : 0)
                - (perfectLoop ? PERFECT_LOOP_BONUS : 0);

        assert !Double.isNaN(fitness) :
                "Invalid fitness  isLoop=" + isLoop + " consistentLoop=" + consistentLoop
                + " numTiles=" + numTiles + " distance=" + distance;
        return Math.max(0, fitness);
    }

    /**
     * First add all the tiles to a hash keyed on location.
     * Then for every one of the six sides of each tile, add one if the
     * neighbor is in the hash. Return (num nbrs in hash - 2(numTiles-1))/numTiles
     * @param path the path to determine compactness of.
     * @return measure of path compactness between 0 and ~1
     */
    private double determineCompactness(TantrixPath path) {
        Set<Location> locationHash = new HashSet<>();
        int numTiles = path.size();
        for (TilePlacement p : path.getTilePlacements()) {
            locationHash.add(p.getLocation());
        }
        int ct = 0;
        for (TilePlacement p : path.getTilePlacements()) {
            for (int i = 0; i < HexTile.NUM_SIDES; i++)   {
                if (locationHash.contains(HexUtil.getNeighborLocation(p.getLocation(), i))) {
                    ct++;
                }
            }
        }
        return (ct - 2.0 * (numTiles-1)) / numTiles * 0.5;
    }
}
